unmarshal.go

```diff
@@ -0,0 +1,336 @@
+// Fields of the struct can be:
+//
+// - Required field of type multicodec.Multicodec with type annotation
+//   `ipld:"multihash"`
+// - Optional string field containing the link name (last item from the path
+//   within the node). it must contain the annotation `ipld:"name"`
+// - Extra fields to match extra keys in IPLD Node. The type must be compatible
+//   and the type annotation must be `ipld:"key:<extra key>"`
+//
+// Fields must be publicly accessible.
+func ReadLinks(r stream.NodeReader, reslinks interface{}) error {
+	rv := reflect.ValueOf(reslinks)
+
+	if rv.Kind() != reflect.Ptr {
+		return fmt.Errorf("Expected slice pointer to Link type: found non pointer type %s", rv.Type().String())
+	}
+
+	sv := reflect.Indirect(rv)
+	if sv.Kind() != reflect.Slice {
+		return fmt.Errorf("Expected slice pointer to Link type: found non slice type %s", sv.Type().String())
+	}
+
+	lt := sv.Type().Elem()
+	if lt.Kind() != reflect.Struct {
+		return fmt.Errorf("Expected slice pointer to Link type: found non struct Link type %s", lt.String())
+	}
+
+	var f fields
+	f.mapFields(lt, nil)
+
+	if f.link == nil && f.hash == nil {
+		return fmt.Errorf("Expected Link type to have a link field (annotated ipld:\"link\") of type string")
+	}
+
+	var stack []reflect.Value
+	var okStack []bool
+
+	it := stream.Iterate(r, nil)
+	defer it.Close()
+
+	for it.Iter() {
+		if it.TokenType == stream.TokenNode {
+			path := it.StringPath()
+			name := ""
+			if len(path) > 0 {
+				name = path[len(path)-1]
+			}
+			v := reflect.New(lt)
+			if f.name != nil {
+				v.Elem().FieldByIndex(f.name).SetString(name)
+			}
+			if f.spath != nil {
+				v.Elem().FieldByIndex(f.spath).Set(reflect.ValueOf(path))
+			}
+			if f.ipath != nil {
+				v.Elem().FieldByIndex(f.ipath).Set(reflect.ValueOf(it.Path))
+			}
+			stack = append(stack, v)
+			okStack = append(okStack, false)
+
+		} else if it.TokenType == stream.TokenKey && it.Value == links.LinkKey {
+			it.Iter()
+			if f.hash != nil {
+				h, err := ReadLinkPath(it.Value)
+				if err == nil {
+					stack[len(stack)-1].Elem().FieldByIndex(f.hash).Set(reflect.ValueOf(h))
+					okStack[len(okStack)-1] = true
+				}
+			}
+			if f.link != nil {
+				if slink, ok := it.Value.(string); ok {
+					stack[len(stack)-1].Elem().FieldByIndex(f.link).Set(reflect.ValueOf(slink))
+					okStack[len(okStack)-1] = true
+				}
+			}
+
+		} else if key, isKey := it.Value.(string); isKey && it.TokenType == stream.TokenKey {
+			if field, has_field := f.attrs[key]; has_field {
+				it.Iter()
+				setValue(stack[len(stack)-1].Elem().FieldByIndex(field), it)
+			}
+
+		} else if it.TokenType == stream.TokenEndNode {
+			if okStack[len(okStack)-1] {
+				sv.Set(reflect.Append(sv, stack[len(stack)-1].Elem()))
+			}
+			stack = stack[:len(stack)-1]
+			okStack = okStack[:len(okStack)-1]
+
+		}
+	}
+
+	return it.LastError
+}
+
+func Unmarshal(r stream.NodeReader, dest interface{}) error {
```

Comment:
why do we need this whole thing here? do we need a general `Unmarshal` in this package? why cant we just avoid all the reflection mess by forcing: 

```
Unmarshal(in ______, dest ipld.Node) error { ... }
```